File: CA_20_SpecialInfo/bubbleSort.py
#
# bubbleSort.py
# @author bulbasaur
# @description 
# @created 2020-07-29T19:12:54.901Z+08:00
# @last-modified 2020-07-29T19:13:49.306Z+08:00
#

list1 = [89, 38, 11, 78, 96, 44, 19, 25]

for i in range(len(list1)-1):        #循环遍历从0开始到倒数第二个元素下标
    for j in range(len(list1)-1-i):       #循环遍历从0开始，每次循环将最大数放到最末尾
        if list1[j] > list1[j+1]:            #判断两个相邻数的大小，大的往后面排，小的往前面排
            list1[j],list1[j+1] = list1[j+1],list1[j]
            
print(list1)

File: CA_20_SpecialInfo/bubbleSort.c
/**
* bubbleSort.c
* @author bulbasaur
* @description 
* @created 2020-07-29T19:08:48.239Z+08:00
* @copyright None 
* None
* @last-modified 2020-07-29T19:12:11.404Z+08:00
*/

#include<stdio.h>
#include<stdlib.h>
#define N 8

void bubble_sort(int a[],int n);
//***一般实现***
void bubble_sort(int a[],int n)//n为数组a的元素个数
{
    //一定进行N-1轮比较
    for(int i=0; i<n-1; i++)
    {
        //每一轮比较前n-1-i个，即已排序好的最后i个不用比较
        for(int j=0; j<n-1-i; j++)
        {
            if(a[j] > a[j+1])
            {
                int temp = a[j];
                a[j] = a[j+1];
                a[j+1]=temp;
            }
        }
    }
}

//***优化实现***
void bubble_sort_better(int a[],int n)//n为数组a的元素个数
{
    int isSorted = 1;
    //最多进行N-1轮比较
    for(int i=0; i<n-1; i++)
    {
        //每一轮比较前n-1-i个，即已排序好的最后i个不用比较
        for(int j=0; j<n-1-i; j++)
        {
            if(a[j] > a[j+1])
            {
                isSorted = 2;
                int temp = a[j];
                a[j] = a[j+1];
                a[j+1]=temp;
            }
        }
        if(isSorted == 1) break; //如果没有发生交换，说明数组已经排序好了
    }
}
int  main()
{
    int num[N] = {89, 38, 11, 78, 96, 44, 19, 25};
    bubble_sort(num, N); //或者使用bubble_sort_better(num, N);
    
    for(int i=0; i<N; i++) {
        printf("%d  ", num[i]);
    }
    printf("\n");
    
    system("pause");
    return 0;
}
